File: morpheus/controllers/monitor_controller.py
# Copyright (c) 2022-2023, NVIDIA CORPORATION.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import typing
from functools import reduce

import fsspec
from tqdm import tqdm

import cudf

from morpheus.messages import ControlMessage
from morpheus.messages import MessageMeta
from morpheus.messages import MultiMessage
from morpheus.utils.logger import LogLevels
from morpheus.utils.monitor_utils import MorpheusTqdm

logger = logging.getLogger(__name__)


class MonitorController:
    """
    Controls and displays throughput numbers at a specific point in the pipeline.

    Parameters
    ----------
    position: int
        Specifies the monitor's position on the console.
    description : str, default = "Progress"
        Name to show for this Monitor Stage in the console window.
    smoothing : float
        Smoothing parameter to determine how much the throughput should be averaged. 0 = Instantaneous, 1 =
        Average.
    unit : str
        Units to show in the rate value.
    delayed_start : bool
        When delayed_start is enabled, the progress bar will not be shown until the first message is received.
        Otherwise, the progress bar is shown on pipeline startup and will begin timing immediately. In large pipelines,
        this option may be desired to give a more accurate timing.
    determine_count_fn : typing.Callable[[typing.Any], int]
        Custom function for determining the count in a message. Gets called for each message. Allows for
        correct counting of batched and sliced messages.
    log_level : `morpheus.utils.logger.LogLevels`, default = 'INFO'
        Enable this stage when the configured log level is at `log_level` or lower.
    tqdm_class: `tqdm`, default = None
        Custom implementation of tqdm if required.
    """

    controller_count: int = 0

    def __init__(self,
                 position: int,
                 description: str,
                 smoothing: float,
                 unit: str,
                 delayed_start: bool,
                 determine_count_fn: typing.Callable[[typing.Any], int],
                 log_level: LogLevels,
                 tqdm_class: tqdm = None):

        self._progress: tqdm = None
        self._position = position
        self._description = description
        self._smoothing = smoothing
        self._unit = unit
        self._delayed_start = delayed_start
        self._determine_count_fn = determine_count_fn
        self._tqdm_class = tqdm_class if tqdm_class else MorpheusTqdm

        if isinstance(log_level, LogLevels):  # pylint: disable=isinstance-second-argument-not-valid-type
            log_level = log_level.value

        self._log_level = log_level
        self._enabled = None  # defined on first call to _is_enabled

    @property
    def delayed_start(self) -> bool:
        return self._delayed_start

    @property
    def progress(self) -> tqdm:
        return self._progress

    def is_enabled(self) -> bool:
        """
        Returns a boolean indicating whether or not the logger is enabled.
        """

        if self._enabled is None:
            self._enabled = logger.isEnabledFor(self._log_level)

        return self._enabled

    def ensure_progress_bar(self):
        """
        Ensures that the progress bar is initialized and ready for display.
        """

        if (self._progress is None):
            self._progress = self._tqdm_class(desc=self._description,
                                              smoothing=self._smoothing,
                                              dynamic_ncols=True,
                                              unit=(self._unit if self._unit.startswith(" ") else f" {self._unit}"),
                                              mininterval=0.25,
                                              maxinterval=1.0,
                                              miniters=1,
                                              position=self._position)

            self._progress.reset()

    def refresh_progress(self, _):
        """
        Refreshes the progress bar display.
        """
        self._progress.refresh()

    def progress_sink(self, x: typing.Union[cudf.DataFrame, MultiMessage, MessageMeta, ControlMessage, typing.List]):
        """
        Receives a message and determines the count of the message.
        The progress bar is displayed and the progress is updated.

        Parameters
        ----------
        x: typing.Union[cudf.DataFrame, MultiMessage, MessageMeta, ControlMessage, typing.List]
            Message that determines the count of the message

        Returns
        -------
        x: typing.Union[cudf.DataFrame, MultiMessage, MessageMeta, ControlMessage, typing.List]

        """

        # Make sure the progress bar is shown
        self.ensure_progress_bar()

        if (self._determine_count_fn is None):
            self._determine_count_fn = self.auto_count_fn(x)

        # Skip incase we have empty objects
        if (self._determine_count_fn is None):
            return x

        # Do our best to determine the count
        count = self._determine_count_fn(x)

        self._progress.update(n=count)

        return x

    def auto_count_fn(self, x: typing.Union[cudf.DataFrame, MultiMessage, MessageMeta, ControlMessage, typing.List]):
        """
        This is a helper function that is used to determine the count of messages received by the
        monitor.

        Parameters
        ----------
        x: typing.Union[cudf.DataFrame, MultiMessage, MessageMeta, ControlMessage, typing.List]
            Message that determines the count of the message

        Returns
        -------
        Message count.

        """

        # pylint: disable=too-many-return-statements

        if (x is None):
            return None

        # Wait for a list thats not empty
        if (isinstance(x, list) and len(x) == 0):
            return None

        if (isinstance(x, cudf.DataFrame)):
            return lambda y: len(y.index)

        if (isinstance(x, MultiMessage)):
            return lambda y: y.mess_count

        if (isinstance(x, MessageMeta)):
            return lambda y: y.count

        if isinstance(x, ControlMessage):

            def check_df(y):
                df = y.payload().df
                if df is not None:
                    return len(df)

                return 0

            return check_df

        if (isinstance(x, list)):
            item_count_fn = self.auto_count_fn(x[0])
            return lambda y: reduce(lambda sum, z, item_count_fn=item_count_fn: sum + item_count_fn(z), y, 0)

        if (isinstance(x, (str, fsspec.core.OpenFile))):
            return lambda y: 1

        if (hasattr(x, "__len__")):
            return len  # Return len directly (same as `lambda y: len(y)`)

        raise NotImplementedError(f"Unsupported type: {type(x)}")

    def sink_on_completed(self):
        """
        Stops the progress bar and prevents the monitors from writing over each other when the last
        stage completes.
        """

        # Ensure that the progress bar exists even if we dont have any values
        self.ensure_progress_bar()

        # Set the name to complete. This refreshes the display
        self.progress.set_description_str(self.progress.desc + "[Complete]")

        self.progress.stop()

        # To prevent the monitors from writing over eachother, stop the monitor when the last stage completes
        MonitorController.controller_count -= 1

        if (MonitorController.controller_count <= 0 and self._tqdm_class.monitor is not None):
            self._tqdm_class.monitor.exit()
            self._tqdm_class.monitor = None

File: python/morpheus/morpheus/_lib/src/controllers/monitor_controller.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "morpheus/controllers/monitor_controller.hpp"

#include "indicators/setting.hpp"

#include "morpheus/messages/control.hpp"
#include "morpheus/messages/meta.hpp"

#include <cudf/column/column.hpp>
#include <cudf/column/column_factories.hpp>
#include <cudf/table/table.hpp>

#include <optional>
#include <type_traits>
#include <vector>

namespace morpheus {

// Component public implementations
// ****************** MonitorController ************************ //

template <typename MessageT>
MonitorController<MessageT>::MonitorController(const std::string& description,
                                               std::optional<std::function<int(MessageT)>> determin_count_fn) :
  m_description(description),
  m_determine_count_fn(determin_count_fn),
  m_count(0)
{
    m_progress_bar.set_option(indicators::option::BarWidth{50});
    m_progress_bar.set_option(indicators::option::Start{"["});
    m_progress_bar.set_option(indicators::option::Fill("■"));
    m_progress_bar.set_option(indicators::option::Lead(">"));
    m_progress_bar.set_option(indicators::option::Remainder(" "));
    m_progress_bar.set_option(indicators::option::End("]"));
    m_progress_bar.set_option(indicators::option::PostfixText{m_description});
    m_progress_bar.set_option(indicators::option::ForegroundColor{indicators::Color::yellow});
    m_progress_bar.set_option(indicators::option::ShowElapsedTime{true});

    MonitorController::m_progress_bars.push_back(m_progress_bar);
}

template <typename MessageT>
MessageT MonitorController<MessageT>::progress_sink(MessageT msg)
{
    if (m_determine_count_fn == std::nullopt)
    {
        m_determine_count_fn = auto_count_fn(msg);
    }

    m_count += (*m_determine_count_fn)(msg);
    m_progress_bar.set_progress(m_count);

    return msg;
}

template <typename T>
struct is_vector : std::false_type
{};

template <typename T, typename U>
struct is_vector<std::vector<T, U>> : std::true_type
{};

template <typename MessageT>
auto MonitorController<MessageT>::auto_count_fn(MessageT msg) -> std::optional<std::function<int(MessageT)>>
{
    if constexpr (std::is_same_v<MessageT, cudf::table>)
    {
        return [](MessageT msg) {
            return msg.num_rows();
        };
    }

    if constexpr (std::is_same_v<MessageT, MessageMeta>)
    {
        return [](MessageT msg) {
            return msg.count();
        };
    }

    if constexpr (std::is_same_v<MessageT, ControlMessage>)
    {
        return [](MessageT msg) {
            if (!msg.payload())
            {
                return 0;
            }
            return msg.payload()->count();
        };
    }

    if constexpr (is_vector<MessageT>::value)
    {
        if (msg.empty())
        {
            return std::nullopt;
        }

        return [this](MessageT msg) {
            auto item_count_fn = auto_count_fn<MessageT>(msg[0]);
            if (item_count_fn == std::nullopt)
            {
                return 0;
            }
            return std::accumulate(msg.begin(), msg.end(), 0, [item_count_fn](int sum, const auto& item) {
                return sum + (*item_count_fn)(item);
            });
        };
    }

    throw std::runtime_error("Unsupported message type received for MonitorController");
}

template <typename MessageT>
void MonitorController<MessageT>::sink_on_completed()
{
    m_progress_bar.mark_as_completed();
}

}  // namespace morpheus
